### PR TITLE
[SPARK-39872][SQL] Change to use `BytePackerForLong#unpack8Values` with Array input api in `VectorizedDeltaBinaryPackedReader`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr use `BytePackerForLong#unpack8Values` with `Array` input api  instead of `BytePackerForLong#unpack8Values` with `ByteBuffer` input api in `VectorizedDeltaBinaryPackedReader` to improve the performance of scanning `INT` and `BIGINT` type data from Parquet DataPage V2.

### Why are the changes needed?
`BytePackerForLong#unpack8Values` with `ByteBuffer` input api has hotspot in using `HeapByteBuffer#get(int)` due to `HeapByteBuffer#get(int)` needs boundary check every time it is called,  this affects the performance of scanning `INT` and `BIGINT` type data from Parquet DataPage V2.

The results of scenario `SQL Single INT/BIGINT Column Scan` as follows:

**Before**

```
OpenJDK 64-Bit Server VM 1.8.0_322-b06 on Linux 5.13.0-1021-azure
Intel(R) Xeon(R) Platinum 8171M CPU @ 2.60GHz
SQL Single INT Column Scan:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet Vectorized: DataPageV1                  192            203          12         82.0          12.2      93.5X
SQL Parquet Vectorized: DataPageV2                  363            373          10         43.4          23.1      49.5X

OpenJDK 64-Bit Server VM 1.8.0_322-b06 on Linux 5.13.0-1021-azure
Intel(R) Xeon(R) Platinum 8171M CPU @ 2.60GHz
SQL Single BIGINT Column Scan:            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet Vectorized: DataPageV1                  252            262          11         62.3          16.1      92.8X
SQL Parquet Vectorized: DataPageV2                  530            537          10         29.7          33.7      44.2X
```

**After**

```
OpenJDK 64-Bit Server VM 1.8.0_342-b07 on Linux 5.15.0-1014-azure
Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
SQL Single INT Column Scan:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet Vectorized: DataPageV1                  173            201          70         91.1          11.0      91.2X
SQL Parquet Vectorized: DataPageV2                  266            272           9         59.2          16.9      59.2X

OpenJDK 64-Bit Server VM 1.8.0_342-b07 on Linux 5.15.0-1014-azure
Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
SQL Single BIGINT Column Scan:            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet Vectorized: DataPageV1                  235            243           9         66.9          15.0      88.3X
SQL Parquet Vectorized: DataPageV2                  398            404           6         39.5          25.3      52.1X
```

After this pr, for the scanning scenario of type `INT` and `BIGINT` data, the performance gap between Parquet DataPage V2 and Parquet DataPage V1 is reduced.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass GitHub Actions

### CODE REVIEW

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java
Comment (LuciferYang) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:304]: Actually, in current usage way, `buffer.arrayOffset()` should always 0

**Comment by LuciferYang** [comment]
The hotspot as follows:

**Before**

image


**After**

image

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java
Comment (LuciferYang) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:301]: Testing the performance of the following changes:

```
  private void unpackMiniBlock() throws IOException {
    Arrays.fill(this.unpackedValuesBuffer, 0);
    BytePackerForLong packer = Packer.LITTLE_ENDIAN.newBytePackerForLong(
        bitWidths[currentMiniBlock]);
    int bitWidth = packer.getBitWidth();
    int times = miniBlockSizeInValues / 8;
    int total = times * bitWidth;
    ByteBuffer buffer = in.slice(total);
    byte[] array = buffer.array();
    int i = 0;
    int srcOffset = buffer.arrayOffset() + buffer.position();
    int targetOffSet = 0;
    while (i < times) {
      packer.unpack8Values(array, srcOffset, unpackedValuesBuffer, targetOffSet);
      i++;
      srcOffset += bitWidth;
      targetOffSet += 8;
    }
    remainingInMiniBlock = miniBlockSizeInValues;
    currentMiniBlock++;
  }
```

Comment (LuciferYang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:301]: after tests, there is no significant performance difference, keep the minimum change

**Comment by LuciferYang** [comment]
cc @sunchao

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java
Comment (LuciferYang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:301]: `SingleBufferInputStream#slice`  has fewer calls, but there is no significant performance difference

**Comment by LuciferYang** [comment]
friendly ping @sunchao @dongjoon-hyun

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for pinging me, @LuciferYang .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master for Apache Spark 3.4.

**Comment by LuciferYang** [comment]
thanks @dongjoon-hyun

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: I think it's better to check `buffer.hasArray` before calling this, since `parquet-mr` may not necessarily return heap buffer all the time.

Comment (LuciferYang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: @sunchao Double check the code path,  `in` is an `ByteBufferInputStream` and passed in by `initFromPage`  method.

The `ByteBufferInputStream` here should come from

https://github.com/apache/spark/blob/bf8a4c47ac3752edf86f1e14e4050e8d202b34a4/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java#L389

`page.getData().toInputStream()` always return a `SingleBufferInputStream` wrapped `HeapByteBuffer`, `SingleBufferInputStream. slice` should still return a `HeapByteBuffer`, so  in the current code path, `buffer.hasArray` should always return true. If I'm wrong, please correct me.

Comment (sunchao) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: Yes, that's the current state, but it'll subject to change though. I know there are some discussion to make more use of direct byte buffer in `parquet-mr`. Therefore, it'd be better to make this code more future proof.

Comment (sunchao) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: (In fact, we've already changed this internally to return direct byte buffer if some flag is turned on).

Comment (LuciferYang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: Got it, let me think about how to reduce the number of calls to `buffer.hasArray`

Comment (dongjoon-hyun) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: Since it's different from AS-IS code, it would be great if we can add a comment about the official Parquet JIRA issue. Could you give me the pointer, @sunchao ?

Comment (dongjoon-hyun) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: BTW, I believe it's worth to have another independent JIRA, @sunchao and @LuciferYang . Please file a new one and link to Parquet JIRA which @sunchao is going to provide.

Comment (LuciferYang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: > @LuciferYang , please file a new one and link to Parquet JIRA which @sunchao is going to provide.

OK

Comment (sunchao) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: there is no parquet JIRA yet

Comment (dongjoon-hyun) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:303]: Could you file one for trace-ability in order not to forget, @sunchao ?

**Comment by sadikovi** [comment]
I would prefer to not rely on the facts:
- that ByteBufferInputStream has only one byte buffer
- that byte buffer is backed by JVM heap.

It would be good if you could revisit this code change.

**Comment by LuciferYang** [comment]
> I would prefer to not rely on:
> 
> * that ByteBufferInputStream is backed by only one byte buffer
> * that byte buffer is backed by JVM heap.
> 
> It would be good if you could revisit this code change.

If we need to consider these factors that are not based on the current code base, we should roll back this patch

**Comment by LuciferYang** [comment]
I think what you said is also reasonable @sadikovi

**Comment by sadikovi** [comment]
It is fantastic to have performance improvements like this one 👍 but it would also be good to avoid future regressions if possible. People use different parquet-mr versions with Spark so it would be good to either have a constraint/error so they know what went wrong or have some kind of general way of handling things.

I don't think it has anything to do with the current codebase, not the Spark one anyway. It would be up to committers to revert but IMHO, it is too drastic - we can just update the code to make it future-proof, plus the change is in master for now.

For example, we could add another code path for it that is only enabled when certain conditions are met. Also, we could add a feature flag that allows to fall back to the previous version. We can also rewrite the page deserialisation logic instead of relying on parquet-mr machinery, this should make reads even faster.

**Comment by LuciferYang** [comment]
Very good suggestion.

For dependency conditions:

- `that ByteBufferInputStream is backed by only one byte buffer`: I think we don't rely on this because the data is processed after 'slice',  if not, please correct me

- `that byte buffer is backed by JVM heap`: https://github.com/apache/spark/pull/37293#discussion_r938030249 is discussing this now

For `rewrite the page deserialisation logic`,   sorry that I haven't done in-depth research,  do you have any ideas? @sadikovi

**Comment by dongjoon-hyun** [comment]
Thank you, @sadikovi. I believe we are going to have a best way in the end like your suggestions (another code path and flags). Could you make a contribution for that by yourself if you can? 😄 

BTW, I just want to point out some concerning two specific reasoning and perspectives in the discussion thread.

First, in the following comment, may I ask why it is good for for Apache Spark, @sadikovi ? As we know, the unknown downstream forks can do everything except blocking the Apache Spark repository because Apache Spark community is deserved to have the best of bests for the community instead of the worst of the worsts. Apache Spark repository is unable to be responsible for unknown downstream activity if you cannot make a contribution back.

> People use different parquet-mr versions with Spark so it would be good to either have a constraint/error so they know what went wrong or have some kind of general way of handling things.

Second, for the following, unfortunately, I would cast my veto to those kind of reverting PRs whose purpose and benefit is unclear in the community. In general, as a reviewer comment from some contributor to another contributor, the following sentence looks too aggressive and is not the best way. There is a better way to do the same thing. You can ping those committers you want to ping to get more feedbacks. That is welcome and a better Apache Spark Way.
> I don't think it has anything to do with the current codebase, not the Spark one anyway. It would be up to committers to revert

BTW, I fully understand what you meant in general here and again I believe we can able to build the best Apache Spark for the whole community while keeping consistency and minimizing burdens with your downstream forks.

**Comment by sadikovi** [comment]
Yes, I can open a follow-up PR to make the code slightly more robust to internal implementation changes in Parquet-Mr.

-----------

@dongjoon-hyun Respectfully, I don't quite understand what you meant in your post and what was concerning in the discussion. I would appreciate it if you could elaborate but I also tried to clarify some of discussion points below.

I was pointing out the code that was relying on the internal implementation details of a library (see [Hyrum's law](https://www.hyrumslaw.com/)) instead of using public API of such library and was merely suggesting to minimise the dependency.

I was not considering the revert of this PR. In fact, I posted my suggestions and specifically advocated for not reverting the PR as it brings performance improvements in my comment. The sentence you mentioned was a reply to the author's comment stating it had something to do with the current, existing, codebase (instead, in my opinion, it is an example of Hyrum's law with regard to the code change relying on Parquet-Mr library and using internal interfaces). Also, I was expressing my opinion in that sentence as I wrote "I think it ...". I also agreed with the author's response with the 👍 reaction.

It appears you might have taken parts of the sentence out of context to accuse me of being disrespectful to or aggressive towards the author in some way which may be a subject of defamation.

FYI, this PR was reviewed and merged by one person without bringing additional committers or reviewers before the merge. I discovered this PR by reviewing commits in master branch.

I don't have any downstream forks. My statement regarding users of Apache Spark was general and I had provided possible fixes/improvements.

**Comment by dongjoon-hyun** [comment]
If you feel in that way, sorry about that. I must admin that I got a wrong impression on your comment before. :) Thank you for saying that.

**Comment by LuciferYang** [comment]
I'm sorry for the trouble I caused.  @sadikovi @dongjoon-hyun

**Comment by sadikovi** [comment]
It is all good, no worries.

**Pull request closed without merging**
